extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetMetadataFileWriterImpl.java

```diff
@@ -0,0 +1,286 @@
+            this.filePath = filePath;
+            this.metadata = metadata;
+        }
+    }
+
+    private final Path metadataRootDirAbsPath;
+    private final List<ParquetFileMetadata> parquetFileMetadataList;
+    private final SeekableChannelsProvider channelsProvider;
+    private final MessageType partitioningColumnsSchema;
+
+    // The following fields are used to accumulate metadata for all parquet files
+    private MessageType mergedSchema;
+    private String mergedCreatedByString;
+    private final Map<String, String> mergedKeyValueMetaData;
+    private final List<BlockMetaData> mergedBlocks;
+    /**
+     * Per-column type information stored in key-value metadata
+     */
+    private List<ColumnTypeInfo> mergedColumnTypes;
+    private String mergedVersion;
+
+    /**
+     * @param metadataRootDir The root directory for the metadata files
+     * @param destinations The individual parquet file destinations, all of which must be contained in the metadata root
+     * @param partitioningColumnsSchema The common schema for partitioning columns to be included in the
+     *        {@value ParquetUtils#COMMON_METADATA_FILE_NAME} file, can be null if there are no partitioning columns.
+     */
+    ParquetMetadataFileWriterImpl(@NotNull final File metadataRootDir, @NotNull final File[] destinations,
+            @Nullable final MessageType partitioningColumnsSchema) {
+        if (destinations.length == 0) {
+            throw new IllegalArgumentException("No destinations provided");
+        }
+        this.metadataRootDirAbsPath = metadataRootDir.getAbsoluteFile().toPath();
+        final String metadataRootDirAbsPathString = metadataRootDirAbsPath.toString();
+        final File firstDestination = destinations[0];
+        for (int i = 0; i < destinations.length; i++) {
+            final File destination = destinations[i];
+            if (!destination.getAbsolutePath().startsWith(metadataRootDirAbsPathString)) {
+                throw new UncheckedDeephavenException("All destinations must be nested under the provided metadata root"
+                        + " directory, provided destination " + destination.getAbsolutePath() + " is not under " +
+                        metadataRootDirAbsPathString);
+            }
+            // TODO How should I change the basename in the API for writeKeyValuePartitioned data for this check?
+            if (i > 0) {
+                // We use filename to generate the key for each file's metadata in the common metadata file, therefore
+                // all files must have unique names.
+                if (destination.getName().equals(firstDestination.getName())) {
+                    throw new UncheckedDeephavenException("When generating common metadata for multiple parquet files, "
+                            + "all files must have unique names, but " + destination.getName() + " is repeated.");
+                }
+            }
+        }
```

Comment:
I have added this check here for unique names, but now I think its better to just enforce a {uuid} in the file name for partitioned writing. Do you have any suggestions for this?